src/legacy/core_plugins/console/public/np_ready/application/models/sense_editor/sense_editor.ts

```diff
@@ -0,0 +1,512 @@
+  }
+
+  prevRequestStart = (rowOrPos?: number | Position): Position => {
+    let curRow: number;
+
+    if (rowOrPos == null) {
+      curRow = this.coreEditor.getCurrentPosition().lineNumber;
+    } else if (_.isObject(rowOrPos)) {
+      curRow = (rowOrPos as Position).lineNumber;
+    } else {
+      curRow = rowOrPos as number;
+    }
+
+    while (curRow > 0 && !this.parser.isStartRequestRow(curRow, this.coreEditor)) curRow--;
+
+    return {
+      lineNumber: curRow,
+      column: 1,
+    };
+  };
+
+  nextRequestStart = (rowOrPos?: number | Position) => {
+    let curRow: number;
+    if (rowOrPos == null) {
+      curRow = this.coreEditor.getCurrentPosition().lineNumber;
+    } else if (_.isObject(rowOrPos)) {
+      curRow = (rowOrPos as Position).lineNumber;
+    } else {
+      curRow = rowOrPos as number;
+    }
+    const maxLines = this.coreEditor.getValue().split('\n').length;
+    for (; curRow < maxLines - 1; curRow++) {
+      if (this.parser.isStartRequestRow(curRow, this.coreEditor)) {
+        break;
+      }
+    }
+    return {
+      row: curRow,
+      column: 0,
+    };
+  };
+
+  autoIndent = _.debounce(async () => {
+    await this.coreEditor.waitForLatestTokens();
+    const reqRange = await this.getRequestRange();
+    if (!reqRange) {
+      return;
+    }
+    const parsedReq = await this.getRequest();
+
+    if (!parsedReq) {
+      return;
+    }
+
+    if (parsedReq.data && parsedReq.data.length > 0) {
+      let indent = parsedReq.data.length === 1; // unindent multi docs by default
+      let formattedData = utils.reformatData(parsedReq.data, indent);
+      if (!formattedData.changed) {
+        // toggle.
+        indent = !indent;
+        formattedData = utils.reformatData(parsedReq.data, indent);
+      }
+      parsedReq.data = formattedData.data;
+
+      this.replaceRequestRange(parsedReq, reqRange);
+    }
+  }, 25);
+
+  update = async (data: string, reTokenizeAll = false) => {
+    return this.coreEditor.setValue(data, reTokenizeAll);
+  };
+
+  replaceRequestRange = (newRequest: any, requestRange: Range) => {
+    const text = utils.textFromRequest(newRequest);
+    if (requestRange) {
+      this.coreEditor.replaceRange(requestRange, text);
+    } else {
+      // just insert where we are
+      this.coreEditor.insert(this.coreEditor.getCurrentPosition(), text);
+    }
+  };
+
+  getRequestRange = async (lineNumber?: number): Promise<Range | null> => {
+    await this.coreEditor.waitForLatestTokens();
+
+    if (this.parser.isInBetweenRequestsRow(lineNumber)) {
+      return null;
+    }
+
+    const reqStart = this.prevRequestStart(lineNumber);
+    const reqEnd = this.nextRequestEnd(reqStart);
+
+    return {
+      start: {
+        ...reqStart,
+      },
+      end: {
+        ...reqEnd,
+      },
+    };
+  };
+
+  getEngulfingRequestsRange = async (
+    range = this.coreEditor.getSelectionRange()
+  ): Promise<Range | null> => {
+    // if (_.isUndefined(cb)) {
```

Comment:
Do we want to keep this commented-out code? Same for the code on ln 169.

Reply:
Will clean this up! Some of this was pre-existing but can all be removed :+1: